src/Microsoft.Data.DataView/ColumnType.cs

```diff
@@ -0,0 +1,374 @@
+        /// </remarks>
+        public Type RawType { get; }
+
+        // IEquatable<T> interface recommends also to override base class implementations of
+        // Object.Equals(Object) and GetHashCode. In classes below where Equals(ColumnType other)
+        // is effectively a referencial comparison, there is no need to override base class implementations
+        // of Object.Equals(Object) (and GetHashCode) since its also a referencial comparison.
+        public abstract bool Equals(ColumnType other);
+    }
+
+    /// <summary>
+    /// The abstract base class for all non-primitive types.
+    /// </summary>
+    public abstract class StructuredType : ColumnType
+    {
+        protected StructuredType(Type rawType)
+            : base(rawType)
+        {
+        }
+    }
+
+    /// <summary>
+    /// The abstract base class for all primitive types. Values of these types can be freely copied
+    /// without concern for ownership, mutation, or disposing.
+    /// </summary>
+    public abstract class PrimitiveType : ColumnType
+    {
+        protected PrimitiveType(Type rawType)
+            : base(rawType)
+        {
+            if (typeof(IDisposable).GetTypeInfo().IsAssignableFrom(RawType.GetTypeInfo()))
+                throw new ArgumentException("A " + nameof(PrimitiveType) + " cannot have a disposable " + nameof(RawType), nameof(rawType));
+        }
+    }
+
+    /// <summary>
+    /// The standard text type.
+    /// </summary>
+    public sealed class TextType : PrimitiveType
+    {
+        private static volatile TextType _instance;
+        public static TextType Instance
+        {
+            get
+            {
+                if (_instance == null)
+                    Interlocked.CompareExchange(ref _instance, new TextType(), null);
+                return _instance;
+            }
+        }
+
+        private TextType()
+            : base(typeof(ReadOnlyMemory<char>))
+        {
+        }
+
+        public override bool Equals(ColumnType other)
+        {
+            if (other == this)
+                return true;
+            Debug.Assert(!(other is TextType));
+            return false;
+        }
+
+        public override string ToString() => "Text";
+    }
+
+    /// <summary>
+    /// The standard number types.
+    /// </summary>
+    public sealed class NumberType : PrimitiveType
+    {
+        private readonly string _name;
+
+        private NumberType(Type rawType, string name)
+            : base(rawType)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(name));
+            _name = name;
+        }
+
+        private static volatile NumberType _instI1;
+        public static NumberType I1
+        {
+            get
+            {
+                if (_instI1 == null)
+                    Interlocked.CompareExchange(ref _instI1, new NumberType(typeof(sbyte), "I1"), null);
+                return _instI1;
```

Comment:
It's unlikely to make a real difference, especially on x86/64, but this will result in two volatile reads instead of one, since you're reading from the volatile first for the null check and then for the value to return.  You can instead do something like:
```C#
NumberType result = _instI1;
if (result == null)
{
    Interlocked.CompareExchange(ref _instI1, new NumberType(typeof(sbyte), "I1"), null);
    result = _instI1;
}
return result;
```
or if you wanted to be fancy about it:
```C#
return
    _instI1 ??
    Interlocked.CompareExchange(ref _instI1, new NumberType(typeof(sbyte), "I1"), null) ??
    _instI1;
```

Reply:
Good call.  Addressed in #2296.